File: 3.x.x/crates-utopia/emails/src/validator/email_local.rs
use serde_json::Value;
use utopia_validators::{Validator, ValueType};

use crate::Email;

/// Validate that an email has a valid local part (PHP `EmailLocal`).
#[derive(Debug, Clone, Default)]
pub struct EmailLocal;

impl EmailLocal {
    /// Create a local-part validator.
    pub fn new() -> Self {
        Self
    }
}

impl Validator for EmailLocal {
    fn description(&self) -> String {
        "Value must be a valid email address with a valid local part".into()
    }

    fn is_array(&self) -> bool {
        false
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, value: &Value) -> bool {
        let Some(raw) = value.as_str() else {
            return false;
        };
        Email::new(raw)
            .map(|email| email.is_valid() && email.has_valid_local())
            .unwrap_or(false)
    }
}

File: packages/emails/src/Emails/Validator/EmailLocal.php
<?php

namespace Utopia\Emails\Validator;

use Utopia\Emails\Email;
use Utopia\Validator;

/**
 * EmailLocal
 *
 * Validate that an email address has a valid local part
 */
class EmailLocal extends Validator
{
    /**
     * Get Description
     *
     * Returns validator description
     */
    public function getDescription(): string
    {
        return 'Value must be a valid email address with a valid local part';
    }

    /**
     * Is valid
     *
     * Validation will pass when $value is a valid email address with a valid local part
     *
     * @param  mixed  $value
     */
    public function isValid($value): bool
    {
        if (! is_string($value)) {
            return false;
        }

        try {
            $email = new Email($value);

            return $email->isValid() && $email->hasValidLocal();
        } catch (\Exception $e) {
            return false;
        }
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }
}
